### DISCUSSION TITLE
Host on Hop.io Ignite (update: no longer unlimited free)

### DISCUSSION THREAD

**Discussion by ganigeorgiev** [pocketbase/pocketbase#2856] [Show and tell]
**3 Nov 2023 Update: Hop.io recently restructured their pricing and no longer offer unlimited free hosting. For more details see https://hop.io/blog/changelog-5.**

---

_This was suggested to me couple months ago since several users complained that Fly started to require billing details for new accounts, so I decided to give it a try while preparing the benchmarks repo._

[Hop.io free tier](https://hop.io/pricing) VM is similar to the one from Fly with slightly more ram (512MB) and 3GB free disk space without requiring billing details (_at least for now_).

**NB! I have no real experience with Hop and I don't know whether they are trustworthy and reliable in the long run, but for small POCs and noncritical hobby projects it should be fine, considering that free hosting tiers are hard to find nowadays.**

_~From my brief tests, it feels a little more "slugish" compared to Fly, but maybe that's because of the `0.5vCPU` or because the deployment is in [`us-east-1`](https://docs.hop.io/reference/regions#regions) and I'm testing from Bulgaria.~ It turned out that the free tier actually allow 1vCPU, it is just that for the "Tiny" plan it is set by defaut to v0.5vCPU. This can be changed either during the deployment by choosing the "Custom" option or from the Hop dashboard > Resources._

_Testbed and benchmarks repo running on Hetzner CAX11 (€3.79), the free fly.io and the free hop.io tiers will be published sometime later this week._

The below guide will use a local Dockerfile (_make sure to [install Docker](https://docs.docker.com/engine/install/) if you haven't already_), but you can also deploy directly from GitHub. For more details, please check [Hop Deployment methods](https://docs.hop.io/ignite/deploying#deployment-methods).

## 1. Create the PocketBase Dockerfile

Create a new local directory with the following `Dockerfile`:

```dockerfile
FROM alpine:latest

ARG PB_VERSION=0.19.0

RUN apk add --no-cache \
   unzip \
   ca-certificates

# download and unzip PocketBase
ADD https://github.com/pocketbase/pocketbase/releases/download/v${PB_VERSION}/pocketbase_${PB_VERSION}_linux_amd64.zip /tmp/pb.zip
RUN unzip /tmp/pb.zip -d /pb/

# uncomment to copy the local pb_migrations dir into the image
# COPY ./pb_migrations /pb/pb_migrations

# uncomment to copy the local pb_hooks dir into the image
# COPY ./pb_hooks /pb/pb_hooks

EXPOSE 8080

# start PocketBase
CMD ["/pb/pocketbase", "serve", "--http=0.0.0.0:8080"]
```

## 2. Install Hop CLI

Go to [Hop.io](https://hop.io/) and create a free account and project.

Then follow the [official instructions to install Hop CLI](https://docs.hop.io/cli/overview#install).

After that you can finally authenticate with your Hop account by running **`hop auth login`**.

This will open a browser window to login and after that it may ask you to also add the Hop registry to your Docker config.
```sh
# type "y"
> Docker was detected, would you like to login to the Hop registry? [y/N] y
```
_For more details you can check https://docs.hop.io/ignite/images._

## 3. Setup and deploy

Run **`hop deploy`** and answer the CLI questions as follow:

```sh
# type your project name
> Deployment name [hop]: your_project_name

# choose "Custom" to select the max allowed resources for the free tier (1vCPU, 512MB ram)
> Select a tier that will suit you well:
      Tiny - Perfect for super small applications, such as a Discord bot or basic REST API (0.5 CPU, 512MB ram)
      Small - Perfect for lightweight services, such as a small Redis instance (1 CPU, 2048MB ram)
      Medium - Perfect for larger database deployments (2 CPU, 4096MB ram)
      Large - Great for more intensive applications, such as game servers (4 CPU, 8192MB ram)
    > Custom - Customize the tier to your needs

# type "1"
> CPUs: 1

# choose "512M"
> Memory: 512M

# type "y" to attach a persistent volume
> Would you like to attach a volume? [y/N] y

# type Enter for the default free 3GB volume
> Volume size [3G]:

# choose "ext4"
# (xfs is usually recommended when working with large files, so I'm not sure if here will have any noticeable difference if you choose it)
> Filesystem:
    > ext4
      xfs

# use the mount path from your Dockerfile
# (in our case it should be "/pb/pb_data")
> Mount path [/data]: /pb/pb_data

# type "n" (it can be changed later from their dashboard if you need it)
> Add environment variables? [y/N] n

# type "n" (it can be changed later from their dashboard if you need it)
> Do you want to change advanced settings? [y/N] n

# type "y" to confirm
> warn: Free tier has been applied to the resources
    info: Estimated monthly price: 0.00$
    Do you want to continue? [y/n] y

> Deployed successfully, you can find it at: https://console.hop.io/ignite/deployment/...
```

Open the link and click on the **"Rollout"** button in the top-right corner:

![rollout_screenshot](https://i.imgur.com/zt5oJ7w.png)

Then go to **"Gateways"** to create a public url for the container.
![gateways_screenshot](https://i.imgur.com/jR0WBKb.png)

_Type an optional tag/name and the port exposed from the Dockerfile (default to 8080)._

**That's it!** You can navigate to the PocketBase admin dashboard as usual using the created gateway link, eg. `https://demo.hop.sh/_/`.

To redeploy, run **`hop deploy`** again.

For more advanced configurations and details, please refer to the [Hop docs and support channels](https://docs.hop.io/getting-started/introduction).

**Comment by eqoram** [comment]
Did anyone already do some benchmarking with pocketbase on the hop.io free tier? :)

**Reply by ganigeorgiev** [reply]
I'm currently working on this. Will publish the results a little bit later.

**Reply by Kamek437** [reply]
Any more info here?

**Reply by ganigeorgiev** [reply]
Sorry, I forgot to post the repo here. You can check https://github.com/pocketbase/benchmarks.

**Comment by 1l0** [comment]
I have a problem on this. When I redeploy (rollout), the database in the volume would be reset.

**Reply by ganigeorgiev** [reply]
Are you using the correct volume mount path, aka. `/pb/pb_data`?

**Reply by 1l0** [reply]
I am
```
% hop volumes ls -la {deployment id}:/
-rw-r--r--  4096    Jul 13 01:09  data.db
-rw-r--r--  32768   Jul 13 01:09  data.db-shm
-rw-r--r--  107152  Jul 13 01:09  data.db-wal
-rw-r--r--  4096    Jul 13 01:09  logs.db
-rw-r--r--  32768   Jul 13 01:09  logs.db-shm
-rw-r--r--  53592   Jul 13 01:09  logs.db-wal
```

**Reply by ganigeorgiev** [reply]
I don't think you are using the correct volume mount path. If your Dockerfile is the same as in the above guide, then the volume path should be `/pb/pb_data`. If it doesn't work for you, then it could be an issue with hop.io, I'm not sure. You can try contacting support in their [discord channel](https://discord.com/invite/hop).

**Reply by ganigeorgiev** [reply]
You can try deleting your container and your `hop.yml` file and start from scratch using a *different project name (_*I found that Hop has some strange cache issue where deleted rollout containers causes new application with the same name to not work correctly for some reason, but I didn't pursuit further as it could be just me not understanding how to use hop_).

**Reply by 1l0** [reply]
OK you are a lifesaver. With different deployment ID it works like a charm.

**Reply by ganigeorgiev** [reply]
I'll extract some reproducible steps sometime later today and will create an issue for the Hop team to check it as it looks like a bug.

**Reply by ganigeorgiev** [reply]
_@1l0 Actually the above ended up being a result of my confusion due to the state in the Hop UI and it is not a bug. For more details you can check the discussion in https://discord.com/channels/988996693322256426/989212975128801290/1129034943532650496, but the general take is that when you run `hop deploy` on existing deployment then the volume persists, as it is expected. But if you delete a deployment, delete your `hop.yml` and then recreate it, the old volume will be also removed with the deployment deletion, no matter whether you use the same mount path or not for the new one._

**Reply by 1l0** [reply]
I'm confused even now but everything is up and running so it's ok.

**Comment by islamzaoui** [comment]
can I change the region to the EU?

**Reply by ganigeorgiev** [reply]
At the moment it seems that there is only 1 region available `us-east-1` (see [Hop regions](https://docs.hop.io/reference/regions#regions)).

**Comment by islamzaoui** [comment]
I downloaded a backup and now I want to restore it, how i can do that?

**Reply by ganigeorgiev** [reply]
I personally haven't tested it but you can try with the `hop volumes` (run it with the `--help` flag to see all options) commands, and more specifically `hop volumes cp` (I'm not only sure whether directories are allowed to be copied or just a single file).

```sh
hop volumes cp /my/local/path/to/pb_data <your hop deployment name or id>:/

# if the above doesn't allow replacing files, you can try to delete the remote pb_data
# but make sure that you have a local copy of it before running it
hop volumes rm <your hop deployment name or id>:/

# and then try again with cp ...
```

**After the files replacement, you'll have to restart your Hop container in order the new data to take effect!**

If the above doesn't work, you can try to ask in the [Hop discord channel](https://discord.com/invite/hop), maybe there is better way to do it, I'm not sure.

**Reply by ganigeorgiev** [reply]
@IslamZaoui I've just tested the `volumes cp` command and it seems that directories are copied as `archive.zip` instead of transferring each individual file.

I've opened a question in their discord channel - https://discord.com/channels/988996693322256426/989212975128801290/1145258621131817020 and will update the example once a resolution is found.

**Reply by islamzaoui** [reply]
Okey, thx for helping

**Reply by islamzaoui** [reply]
finally, I found a way:

`hop volumes cp ./<name of backup>.zip <your hop deployment name or id>:/backups`

`<name of backup>.zip` is a backup file that I downloaded from pocketbase
Just copy it from local to `<your hop deployment name or id>:/backups>`
This will place the backup file into to pocketbase backups directory
In Pocketbase you will find the backup named backups for some reasons, hit restore and everything is fine

**Reply by ganigeorgiev** [reply]
Yes, that's one possible workaround, but note that it will work only if the backups are locally stored. With the next v0.18/v0.19 release there will be an option to upload a backup archive from the Admin UI that should work with both local and S3 backups (#2599).

**Reply by islamzaoui** [reply]
This is gonna be a great QoL feature, thx again for helping

**Comment by babakfp** [comment]
Hi

Thank you for the new solution ❤️

I have two questions.

1. PocketBase uses port `8090` when running `pocketbase.exe serve`, why do we use port `8080` in the Dockerfile?

2. How can I copy the local `./pb_data` into the Hop.io server? I tried doing this but it didn't work:

```dockerfile
COPY ./pb_data /pb/pb_data
```

It was inspired based on this code from the [above Dockerfile config](https://github.com/pocketbase/pocketbase/discussions/2856#discussion-5370780):

> ```dockerfile
> # uncomment to copy the local pb_migrations dir into the image
> # COPY ./pb_migrations /pb/pb_migrations
>
> # uncomment to copy the local pb_hooks dir into the image
> # COPY ./pb_hooks /pb/pb_hooks
> ```

I also have the `./pb_migrations` folder, but it's also not uploaded to the Hop.io server; I say this because I don't see the folder here (same with `./pb_data` too):

![Screenshot 2023-10-06 233444](https://github.com/pocketbase/pocketbase/assets/44144724/f265c29e-1f55-4e92-8b59-18f5d5b7a3a1)

Thanks

**Reply by ganigeorgiev** [reply]
1) You can use whatever port you want. I think `8080` was just the default of the Hop gateway.

2) It doesn't make sense to copy your local `pb_data` as it will be overwritten with each deploy. If you want to transfer only initially your local data to the remote you can backup your local `pb_data` from the _"Admin UI > Settings > Backups"_ and then upload the backup in your deployed Hop instance again from the PocketBase Admin UI.

> I also have the ./pb_migrations folder, but it's also not uploaded to the Hop.io server;

Did you uncomment the `/pb/pb_migrations` line?
Also note that when copying the pb_migrations in the Dockerfile they will be part of the image not of the volume.
The volume is mounted only for `/pb/pb_data` directory.

**Comment by skyfallwastaken** [comment]
FYI, Hop.io now uses the Railway pricing model - you get a $5 one-time free credit, which doesn't renew every month. This affects existing projects too. https://hop.io/blog/changelog-5

**Reply by ganigeorgiev** [reply]
Thanks for the update. I'll add a note to the changelog in the beginning of the discussion for better visibility.

**Reply by babakfp** [reply]
I'm sad 🥲
I have $4.39 Free credit left 😭
Hop was nice, it was free and faster than what I was using before.
Is it possible to host PocketBase on Vercel somehow? Maybe if you talk to someone in Vercel they like the idea and implement the feature to accept PocketBase apps? Just being optimistic :)

**Reply by skyfallwastaken** [reply]
@babakfp I think that the Google Cloud Free VM has around the same specs as Hop.io - maybe check that out?

**Comment by Saad5400** [comment]
I'm trying to sign up at [console](https://console.hop.io/) but it's asking for an "invite code" ... so how to get one ... ?

![image](https://github.com/pocketbase/pocketbase/assets/86385454/62a8f9d1-aecf-45f0-95b5-82f8d6c6bec3)

**Reply by ganigeorgiev** [reply]
I'm not sure. You can ask in their [discord channel](https://discord.gg/hop) but note that Hop.io has no longer a free plan per https://hop.io/blog/changelog-5.
